### PR TITLE
Restore _AxesStack to track a Figure's Axes order.

### LINKED ISSUE #18598: direct api to add error bars to lineplots and scatterplots
### Problem

Currently error bars can only be drawn through `axes.Axes.errorbar`.
While the existing api is fine, I think the discoverability would be improved by adding support for adding support for `xerr` and `yerr`  kwargs in `axes.Axes.scatter` and `axes.Axes.plot` directly.

### Proposed Solution

Here's what I'd like to work
```python
# -- scatterplot with error bars --
# current
ax.errorbar(data=data, x="x", y="y", xerr="xerr", yerr="yerr", fmt="o")

# proposed
ax.scatter(data=data, x="x", y="y", xerr="xerr", yerr="yerr")

# -- lineplot with error bars --
# current
ax.errorbar(data=data, x="x", y="y", xerr="xerr", yerr="yerr", fmt="-")

# proposed
ax.plot(data=data, x="x", y="y", xerr="xerr", yerr="yerr")
```

### Additional context and prior art

This would improve downstream projects that encapsulate `axes.Axes.plot` and `axes.Axes.scatter`.
For instance `seaborn.regplot` (https://seaborn.pydata.org/generated/seaborn.regplot.html?highlight=regplot#seaborn.regplot) would benefit from such a change and would naturally gain support for user defined error bars through `xerr` and `yerr`.

A simple and maybe naive approach would be to isolate the errorbar functionality from `axes.Axis.errorbar`. I anticipate a side effect would be that a lot of other keyword arguments, specific to `axes.Axis.errorbar` require support in `axes.Axis.plot` and `axes.Axis.scatter`, though hopefully they could be simply passed down through `**kwargs` ?

### LINKED ISSUE #19598: Axes order changed in 3.4.0rc1
### Bug report

**Bug summary**

The order of the `axes` is changed in matplotlib 3.4.0rc1.

**Code for reproduction**

```python
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

image = np.arange(10*10).reshape((10, 10))

fig = plt.figure()
ax = fig.add_subplot()
im = ax.imshow(image)
fig.colorbar(im)

print("Matplotlib version", matplotlib.__version__)
print("Axes:", fig.axes)
```

**Actual outcome**

```
Matplotlib version 3.4.0rc1
axes: [<AxesSubplot:label='<colorbar>'>, <AxesSubplot:>]
```

**Expected outcome**

```
Matplotlib version 3.4.0rc1
axes: [<AxesSubplot:>, <AxesSubplot:label='<colorbar>'>]
```

**Matplotlib version**

  * Operating system: Fedora
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.4.0rc1
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.8.6 or 3.9.2
  * Jupyter version (if applicable): 
  * Other libraries: 

Matplotlib 3.4.0rc1 installed from pypi

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

This is a simplified version of the `AxesStack` removed in #19153, but removing the key entry, and just tracking the index with the `Axes`. We no longer need to track keys.

Fixes #19598.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:100]: probably should test this since it seems someone could change this?

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:111]: What is self._default?  Is it `None`?

Comment (QuLogic) [reply lib/matplotlib/figure.py:111]: Yes, it appears so, since `_AxesStack` doesn't pass it to super init:
https://github.com/matplotlib/matplotlib/blob/bc972942bd41a32c7cd693a3f29a29442d2d6e3f/lib/matplotlib/cbook/__init__.py#L608

File: lib/matplotlib/figure.py
Comment (QuLogic) [reply lib/matplotlib/figure.py:100]: I'll check further, but this may not ever happen, as the `Figure.add_*` might handle that already now.

Comment (QuLogic) [reply lib/matplotlib/figure.py:100]: There are 3 effects of this early `return`:
1. A different return value, but `_AxesStack.add` is only called from `_add_axes_internal`, and that doesn't care about the return value.
2. The `Axes` doesn't get added to the stack multiple times, but that's what we want anyway.
3. The `Axes` would not be made current (i.e., moved to the end), but there's a `sca` in `_add_axes_internal`, so we don't need to worry about that.

I thought some tests checked `add_axes`/`add_subplot` with an existing `Axes`, but it appears not if this isn't getting coverage. I will add some more tests and remove the return value, which no-one uses.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
To summarize?

 -  we used the stack to track three things: what Axes are in the Figure, the order they were added in, and which one is the "current" figure.  When we removed the key-tracking we lost the ability to track the order they were added in inadvertently. 
- Long term it may be simpler to switch to a list and pointer, but this is the lightest touch fix to get back the old behavior.

File: lib/matplotlib/cbook/__init__.py
Comment (lpsinger) [inline lib/matplotlib/cbook/__init__.py:625]: Why is this method removed?

**Review by lpsinger — APPROVED** [review]

File: lib/matplotlib/cbook/__init__.py
Comment (QuLogic) [reply lib/matplotlib/cbook/__init__.py:625]: It was moved from `_AxesStack` to the base class; now that `_AxesStack` is back, it's not needed there. Better not to add more public API if not necessary.

File: lib/matplotlib/figure.py
Comment (timhoffm) [reply lib/matplotlib/figure.py:111]: I'd return `None` explicitly, that's what the docs say, and nobody should overwrite a private attribute, so nobody should ever expect something other than `None`.

**Review by jklymak — APPROVED** [review]

**Comment by jklymak** [comment]
I'll wait until today's meeting to merge in case there are any other issues but I think this is good.

**Comment by QuLogic** [comment]
So we can merge now?

**Comment by jklymak** [comment]
yes 😄

**Comment by QuLogic** [comment]
I don't know why meeseeks is randomly not backporting.

@meeseeksdev backport to v3.4.x

**Pull request merged**
